Implemented Inquisitors

xlenstra opened the pull request:
This PR continues the implementation of religion, by implementing (and nobody expected this) inquisitors.
These religious units have two usages:
- When stationed in or next to a city, no religion can be spread to that city using religious units (natural pressure still works)
- Can 'remove heresy' from your own cities, removing all pressure of other religions than your own (including your former pantheon)

SomeTroglodyte commented:
Well... Soundjay's Terms of use sound nicely formulated, but it's not linked to any of the widely accepted open source related license styles. Wonder if that could be a problem. I'm against the Inquisitor icon, though, as nice as it is - when we get to Spies that might be too close. And the sound is missing the cries of the burning witches...

SomeTroglodyte commented:
How about...
![Inquisitor](https://user-images.githubusercontent.com/63000004/130144406-4ca1341c-3bcf-419f-b6d3-201c9528adac.png)
[Inquisitor.xcf.zip](https://github.com/yairm210/Unciv/files/7017726/Inquisitor.xcf.zip)
... should look without the transparency like: ...
![image](https://user-images.githubusercontent.com/63000004/130144991-2cf5a93c-8610-4240-ad4c-a9fa8686107b.png)
```
[monk][https://thenounproject.com/term/monk/3890596/) by Kevin as part of Inquisitor
[Executioner](https://thenounproject.com/term/inquisition/110009/) by Igor Yanovsky as part of Inquisitor
[Torch](https://thenounproject.com/term/torch/3517926/) by IconMark as part of Inquisitor
```

yairm210 commented:
I veto all images that aren't pure white as unit icons. They get real ugly real fast when tinted in civ colors. Imagine that tinted blue. Or purple.

yairm210 commented on android/assets/jsons/Civ V - Vanilla/Units.json:
```diff
@@ -1532,6 +1532,21 @@
+		"movement": 3,
+		"religiousStrength": 1000,
+		"civilopediaText": [
+			{"text": "Can remove foreign religions from your own cities."}
```
If you need text to explain what the unique enables, it's probably not user friendly enough.

SomeTroglodyte commented:
> Or purple

Yay! No seriously, that's why the xcf is included. Exporting without that layer is a cinch. Generally - not too extreme?

OK then - for the "Remove Heresy" sound... A zip with a tentative mp3 offering and the above png in white only:
[Inquisition.zip](https://github.com/yairm210/Unciv/files/7018071/Inquisition.zip)

Do ***not*** just include without listening, might really be too extreme. But I like it. 2.5 full seconds, though. I'll list sources if and when it's requested.

AdityaMH commented:
I think just using this with some cut his neck.
https://thenounproject.com/term/inquisition/110007/

AdityaMH commented:
> How about...
> ![Inquisitor](https://user-images.githubusercontent.com/63000004/130144406-4ca1341c-3bcf-419f-b6d3-201c9528adac.png)
> [Inquisitor.xcf.zip](https://github.com/yairm210/Unciv/files/7017726/Inquisitor.xcf.zip)
> ... should look without the transparency like: ...
> ![image](https://user-images.githubusercontent.com/63000004/130144991-2cf5a93c-8610-4240-ad4c-a9fa8686107b.png)
> 
> ```
> [monk][https://thenounproject.com/term/monk/3890596/) by Kevin as part of Inquisitor
> [Executioner](https://thenounproject.com/term/inquisition/110009/) by Igor Yanovsky as part of Inquisitor
> [Torch](https://thenounproject.com/term/torch/3517926/) by IconMark as part of Inquisitor
> ```

No doubt, using this will look ugly if set this as icon.
better just head icon much better.

ravignir commented:
What about one of these?:
https://thenounproject.com/term/invisibility-cloak/1419648/

https://thenounproject.com/term/skill-sword-flame/2360212/

AdityaMH commented:
> What about one of these?:
> https://thenounproject.com/term/invisibility-cloak/1419648/

I recommend this, without any editing

logicminimal commented:
> https://thenounproject.com/term/skill-sword-flame/2360212/

i actually like this if this is going to be the icon for remove heresy, could be better if rotated at a 45d angle though

yairm210 commented:
Way to much double hardcoding
"Can [action] [amount] times" as unique
That way, almost everything is generic except for the effect itself
And we're also better equipped for future units with x actions

xlenstra commented:
Fixed the hardcoding problems, also fixed the following bugs:
- If the majority religion in a city was a pantheon religion, no icon would show up in the religious overview
- If a barbarian destroyed a civilian unit, you would not get a notification for this

xlenstra commented:
Speaking about how horrible it is to have two game atlasses: we have reached that point once again. The two images added in this PR overfill the first atlas once again.

Looking at the atlas, however, it appears that all the stat icons are duplicated in it? There are a large versions at the edges and smaller versions more in the middle and spread out

yairm210 commented:
Emoji icons are A. tiny and B. needed for fonts.
Order of what to remove discussed here:
https://github.com/yairm210/Unciv/pull/4954

yairm210 commented on android/assets/jsons/Civ V - Vanilla/Units.json:
```diff
@@ -1528,11 +1528,25 @@
+			"[+1] Visibility Range", "Hidden when religion is disabled", "Unbuildable", "Religious Unit"
+		],
+		"movement": 3,
+		"civilopediaText": [
+			{"text": "Can remove foreign religions from your own cities."}
```
Remove this, the unique is now self explanatory :)

yairm210 commented on core/src/com/unciv/logic/civilization/CivilizationInfo.kt:
```diff
@@ -803,8 +804,10 @@ class CivilizationInfo {
                 else religionManager.religion?.name
-            if (placedUnit.hasUnique("Can spread religion [] times"))
-                placedUnit.abilityUsedCount["Religion Spread"] = 0
+            if (placedUnit.canDoReligiousAction(Constants.spreadReligionAbilityCount))
+                placedUnit.abilityUsedCount[Constants.spreadReligionAbilityCount] = 0
+            if (placedUnit.canDoReligiousAction(Constants.removeHeresyAbilityCount))
+                placedUnit.abilityUsedCount[Constants.removeHeresyAbilityCount] = 0
```
for(unique in placedUnit.getUniques("Can [] [] times) placedUnit.abilityUsedCount[unique.params[0]] = 0

yairm210 commented on core/src/com/unciv/logic/civilization/ReligionManager.kt:
```diff
@@ -125,7 +126,8 @@ class ReligionManager {
         if (religion == null) return false // First found a pantheon
         if (religion!!.isMajorReligion()) return false // Already created a major religion
-        if (prophet.abilityUsedCount["Religion Spread"] != 0) return false // Already used its power for other things
+        if (prophet.abilityUsedCount[Constants.spreadReligionAbilityCount] != null && prophet.abilityUsedCount[Constants.spreadReligionAbilityCount] != 0) return false // Already used its power for other things
+        if (prophet.abilityUsedCount[Constants.removeHeresyAbilityCount] != null && prophet.abilityUsedCount[Constants.removeHeresyAbilityCount] != 0) return false // Already used its power for other things
```
if (prophet.abilityUsedCount.any{it.value>0} return false

yairm210 commented on core/src/com/unciv/models/ruleset/unit/BaseUnit.kt:
```diff
@@ -330,8 +330,10 @@ class BaseUnit : INamed, INonPerpetualConstruction, ICivilopediaText {
             unit.religion = cityConstructions.cityInfo.religion.getMajorityReligionName()
-            if (unit.canSpreadReligion())
-                unit.abilityUsedCount["Religion Spread"] = 0
+            if (unit.canDoReligiousAction(Constants.spreadReligionAbilityCount))
+                unit.abilityUsedCount[Constants.spreadReligionAbilityCount] = 0
+            if (unit.canDoReligiousAction(Constants.removeHeresyAbilityCount))
+                unit.abilityUsedCount[Constants.removeHeresyAbilityCount] = 0
         }
```
Also appears in CivilizationInfo.addUnit, deduplicate

yairm210 commented on core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt:
```diff
@@ -457,37 +459,76 @@ object UnitActions {
+            action = {
+                unit.abilityUsedCount[Constants.removeHeresyAbilityCount] = unit.abilityUsedCount[Constants.removeHeresyAbilityCount]!! + 1
+                city.religion.removeAllPressuresExceptFor(unit.religion!!)
+                unit.currentMovement = 0f
+                if (unit.abilityUsedCount[Constants.removeHeresyAbilityCount] == maxHerseyUses) {
+                    if (unit.isGreatPerson())
+                        addGoldPerGreatPersonUsage(unit.civInfo)
+                    unit.destroy()
+                }
```
This part (plus the "add 1 to uses") would be the same for all actions with X uses. Maybe should be a function, addUse(unit, ability).

yairm210 commented on core/src/com/unciv/ui/worldscreen/unit/UnitTable.kt:
```diff
@@ -163,9 +164,14 @@ class UnitTable(val worldScreen: WorldScreen) : Table(){
                 
-                if (unit.canSpreadReligion()) {
+                if (unit.canDoReligiousAction(Constants.spreadReligionAbilityCount)) {
                     unitDescriptionTable.add(ImageGetter.getStatIcon("Faith")).size(20f)
-                    unitDescriptionTable.add(unit.getReligionString())
+                    unitDescriptionTable.add(unit.getActionString(Constants.spreadReligionAbilityCount))
+                }
+                
+                if (unit.canDoReligiousAction(Constants.removeHeresyAbilityCount)) {
+                    unitDescriptionTable.add(ImageGetter.getImage("OtherIcons/Remove Heresy")).size(20f)
+                    unitDescriptionTable.add(unit.getActionString(Constants.removeHeresyAbilityCount))
```
Hmmm this could also be unified but maybe we'll leave that until the third unit with X uses comes along.

SomeTroglodyte commented:
> Speaking about how horrible it is to have two game atlasses: we have reached that point once again. The two images added in this PR overfill the first atlas.

See also #4959

> Edit: Turns out we have a folder 'emoji-icons' which duplicates all the stat icons, would it help to remove those and just use the stat icons instead? Or is this required for usage in fonts?

Yea, kinda. The bitmaps going into fonts it says in previous comments should be 50x50, and they're baseline-tweaked so they don't dance too badly out of a text row - not necessarily centered in their box. Could possibly be done in code - but - nah no fun.
